libslide/src/scanner.rs

```diff
@@ -0,0 +1,233 @@
+    fn iterate_digit(&mut self,mut i: usize,mut c: char) -> (Token, usize){
+        let mut int_str = "".to_owned();
+        let mut dec_str = "0.".to_owned();
+        let ret: Token;
+        // iterate through integer part
+        while c.is_digit(10) && i < self.input.chars().count() {
+            int_str.push(c);
+            i += 1;
+            if i < self.input.chars().count() {
+                c = self.input.as_bytes()[i] as char;
+            }
+        }
+        // iterate through decimal
+        if{c == '.'}{
+            i += 1;
+            c = self.input.as_bytes()[i] as char;
+            while c.is_digit(10) && i < self.input.chars().count(){
+                dec_str.push(c);
+                i += 1;
+                if i < self.input.chars().count() {
+                    c = self.input.as_bytes()[i] as char;
+                }
+            }
+            // turn integer and decmial strings into token
+            ret = Token{token: TokenType::Num, integer: int_str.parse::<i64>().unwrap(), float: dec_str.parse::<f64>().unwrap()}
+        }
+        else{
+            // turn integer string into token and default the float
+            ret = Token{token:TokenType::Int, integer: int_str.parse::<i64>().unwrap(), ..Default::default()}
+        }
+        return (ret, i);
+    }
+
+    pub fn scan(&mut self){
+        let mut i: usize = 0;
+        let mut c: char;
+        let mut t: Token;
+        let mut tuple: (Token, usize);
+        // iterate through string
+        while i < self.input.chars().count() {
+            c = self.input.as_bytes()[i] as char;
+            // ignore whitespace
+            if !c.is_whitespace() {
+                // check for digit and call correct helper function 
+                if c.is_digit(10) {
+                    tuple = Scanner::iterate_digit(self, i, c);
+                    i = tuple.1;
+                    self.output.push(tuple.0);
+                }
+                else{
+                    t = Scanner::create_symbol_token(c);
+                    if !t.is_empty() {
+                        self.output.push(t);
+                    }
+                    // throw error if token is not correct
+                    else{
+                        panic!("Character invalid");
+                    }
+                    i += 1;
+                }       
+            }
+            else{
+                i += 1;
+            }
+        }
+    }
+    
+}
+
+#[cfg(test)]
+mod tests{
+    use super::*;
+
+    fn compare_vec<T: PartialEq>(a: &Vec<T>, b: &Vec<T>) -> bool{
+        let equal = a.iter().zip(b.iter()).filter(|&(a,b)| a == b).count();
+        equal == a.len() && equal == b.len()
+    }
+
+    #[test]
+    fn test_empty_string_scan() {
+        let mut s = Scanner::new("");
+        s.scan();
+        let result = Vec::<Token>::new();
+        assert_eq!(true, compare_vec::<Token>(&result, &s.output));
+    }
```

Comment:
These tests look good; I will probably replace them after this PR because I have been doing some research on alternative testing strategies that I think may be better suited for this kind of thing. But I will send a PR for that.

Reply:
ok